File: Team5/YaminiSingh/spiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
      int m = matrix.size(), n = m ? matrix[0].size() : 0, ru = 0, rd = m - 1, cl = 0, cr = n - 1, p = 0;
        vector<int> order(m * n);
        while (ru <= rd && cl <= cr) {
            for (int col = cl; col <= cr; col++) {
                order[p++] = matrix[ru][col];
            }
            if (++ru > rd) {
                break;
            }
            for (int row = ru; row <= rd; row++) {
                order[p++] = matrix[row][cr];
            }
            if (--cr < cl) {
                break;
            }
            for (int col = cr; col >= cl; col--) {
                order[p++] = matrix[rd][col];
            }
            if (--rd < ru) {
                break;
            }
            for (int row = rd; row >= ru; row--) {
                order[p++] = matrix[row][cl];
            }
            if (cl++ > cr) {
                break;
            }
        }
        return order;
    }
};

File: Team1/TitikshaSharma/SpiralMatrix.java

//Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> result = new ArrayList<Integer>();
        if (matrix.length == 0)
            return result;
        int maxcn = matrix.length * matrix[0].length;
        int count = 0;
        int row = 0;
        int col = 0;
        int minr = 0;
        int minc = 0;
        int maxr = matrix.length;
        int maxc =  matrix[0].length;
        while (count < maxcn ){
            
            while(col < maxc ){
                result.add(matrix[row][col++]);
                count++;
            }
            col--;
            minr++;
            row++;
            while (row < maxr && count < maxcn){
                result.add(matrix[row++][col]);
                  count++;
            }
            row--;
            maxc--;
            col--;
            while (col >= minc && count < maxcn){
                 result.add(matrix[row][col--]);
                  count++;
            }
            maxr--;
            col++;
            row--;
            while(row >= minr && count < maxcn){
               result.add(matrix[row--][col]);
                  count++;
            }
            row++;
            minc++;
            col++;
        }
         
        return result;
        
        
    }
}
